api/src/main/java/org/apache/iceberg/expressions/Literals.java

```diff
@@ -501,18 +531,22 @@ public <T> Literal<T> to(Type type) {
           return (Literal<T>) new TimeLiteral(timeMicros);
 
         case TIMESTAMP:
-          if (((Types.TimestampType) type).shouldAdjustToUTC()) {
-            long timestampMicros =
-                ChronoUnit.MICROS.between(
-                    EPOCH, OffsetDateTime.parse(value(), DateTimeFormatter.ISO_DATE_TIME));
-            return (Literal<T>) new TimestampLiteral(timestampMicros);
+          TimestampType tsType = (TimestampType) type;
+          if (tsType.shouldAdjustToUTC()) {
+            long timestampUnits =
+                tsType
+                    .unit()
+                    .between(EPOCH, OffsetDateTime.parse(value(), DateTimeFormatter.ISO_DATE_TIME));
```

Comment:
Should this logic move into `DateTimeUtil` now that there are two versions? I think we've avoided that up until now to to avoid modifying this code, but if it is changing anyway we could at least standardize.

Reply:
Maybe. Do you want that done here, or in a follow-up pull request?

Reply:
Yes, I think it should be here.

Reply:
My bad; I thought I knew what you meant when I replied two weeks ago, but I do not! Can you clarify exactly what you'd like to see moved to DateTimeUtil?

At first I thought you might have wanted this whole case moved to DateTimeUtil, but then I realized that would create a circular dependency between Literals and DateTimeUtil.

I reorganized this a bit and maybe it helps, but I still don't know what ought to move out.

Thanks!

Reply:
The conversion logic should be in `DateTimeUtil` rather than embedded here. There may be methods for doing this already for microseconds. The important thing is that we want to centralize the logic in `DateTimeUtil` if we are changing this section. We hadn't done it yet because this code is so old and we saw no reason to modify it.

Reply:
I took another stab at this yesterday.

Please take another look.

Thanks!